### DISCUSSION TITLE
referential integrity checks?

### DISCUSSION THREAD

**Discussion by tmtmtmtm** [dathere/qsv#1872] [Q&A] [answered]
If I have one CSV that includes a `jurisdiction` column, is there currently a way to `validate` that all the values in that column have a matching entry in a `jurisdictions.csv` file? 

I can do that as a one-off check with `qsv join --left-anti`, but I'm not familiar enough with JSON Schema to be able to work out if `qsv validate` can handle this sort of integrity checking.

**Comment by a5dur** [comment]
```qsv join --left-anti``` should do the job, i'm not sure if ```qsv validate``` (which uses JSON schema) is suitable for this (value based) type of cross-file validation, it can still be useful for validating the structure and content of individual CSV files. 

You can use it to ensure that all fields in the csv file are of the correct type, within specified ranges, or match certain patterns.


Although you can try creating a JSON Schema file from jurisdictions.csv file using the ```qsv schema``` command. you might need to fine tune it with the jurisdiction values

See https://github.com/jqnatividad/qsv/blob/master/src/cmd/validate.rs#L2
and https://json-schema.org/draft/2020-12/json-schema-validation.html

**Comment by jqnatividad** [comment]
The `validate` command is designed for high speed validation -  if the `jurisdictions` domain is fairly static, encoding them as enums in JSONschema is the way to go.

Just start with a "training" CSV to run the `schema` command against to produce the starter JSONschema and modify it accordingly. You can then use the `validate` command against CSVs with an identical schema.

https://github.com/jqnatividad/qsv/blob/1efa6b537e986d5d0252fe127c52715a0a4d0645/scripts/NYC_311_SR_2010-2020-sample-1M.csv.schema.json#L30-L67

If the `jurisdictions` domain is dynamic, you may want to look into using the `luau` command's lookup table functions.

I have an extended example at https://github.com/dathere/qsv-lookup-tables.

**Reply by tmtmtmtm** [reply]
Thanks for the examples. What I think I'm looking for is a way for the enum to be a little more dynamic, e.g pointing at a column of another CSV, rather than a hard-coded list in the JSON. If that's not possible, then I guess I could always dynamically update the schema with those values before calling `qsv validate` using something like:

```sh
jq --argjson vals "$(
  qsv select id jurisdictions.csv | qsv behead | qsv fmt --quote-always | jq -s .
)" '.properties.jurisdiction |= $vals' schema.json | sponge schema.json
```

The luau lookup table approach seems intriguing, but that sales tax example, whilst really useful to show some of the more involved possibilities, makes it quite hard for me to imagine what that could look like for a much simpler version. How could that be used to verify that the State column in `data.csv` matched an Abbreviation from the states table?

**Comment by jqnatividad** [comment]
Now that the `jsonschema` crate supports [custom keywords](https://github.com/Stranger6667/jsonschema-rs?tab=readme-ov-file#custom-keywords), I'm considering adding a `dynlookup` keyword to allow validating a field against a CSV - with the CSV being either remote (http/s scheme supported) or in the local filesystem.

Though I'm sure we can have a full-blown `luau` script to do custom validation, it's not as convenient as JSONschema validation.

WDYT?

**Reply by tmtmtmtm** [reply]
Oh, that sounds fantastic, and would be very useful!

**Reply by jqnatividad** [reply, marked as answer]
`validate` now supports the `dynenum` custom keyword for validating against a local CSV or on a URL (http/https schemes supported). 😄 

https://github.com/jqnatividad/qsv/pull/2166

For an example, see:

https://github.com/jqnatividad/qsv/blob/8ded63a0356c92b73fcce786d7364cf5bab7ebaa/scripts/benchmark_data-dynenum.csv.schema.json#L32-L40

https://github.com/jqnatividad/qsv/blob/master/scripts/NYC_agencies.csv

Contrast that with using hardcoded enums:

https://github.com/jqnatividad/qsv/blob/8ded63a0356c92b73fcce786d7364cf5bab7ebaa/scripts/NYC_311_SR_2010-2020-sample-1M.csv.schema.json#L30-L67
